clang_format.py

```diff
@@ -116,29 +166,60 @@ def clang_format_path() -> Path:
     return clang_format_file
 
 
-def main():
+def main(argv: Optional[Sequence[str]] = None) -> Union[int, None]:
+    parser = argparse.ArgumentParser(description="Arguments for pre commit.")
+    parser.add_argument(
+        "version",
+        choices=(
+            "3.5.0",
+            "3.6.0",
+            "3.7.0",
+            "3.9.0",
+            "4.0.0",
+            "5.0.0",
+            "8.0.0",
+            "11.0.0",
+        ),
+        help="Clang format version to run",
+    )
+    parser.add_argument("scope", choices=["diff", "whole-file"], help="Run on files")
+    parser.add_argument(
+        "files",
+        nargs="*",
+        type=Path,
+        help="Files pre-commit believes are changed.",
+    )
+    try:
+        args = parser.parse_args(argv)
+    except SystemExit as e:
+        assert e.code is not None
+        return e.code
+
     this_dir = os.path.dirname(__file__)
     git_cf_path = os.path.join(this_dir, "git-clang-format")
-
-    if sys.argv[1] == "--diff":
-        print("Formatting changed lines in " + " ".join(sys.argv[2:]))
+    if args.scope == "diff":
+        print("Formatting changed lines in " + " ".join(f"{args.files}"))
         subprocess.run(
-            [
+            (
                 sys.executable,
```

Comment:
while we're here (or in a separate commit), should we change these to check_subprocess?  If clang-format fails (e.g. bad .clang-format file) we currently silently swallow that failure.

Reply:
[`check_call`](https://docs.python.org/3/library/subprocess.html#subprocess.check_call)

Reply:
add check=True in the subprocess run